proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1523 @@
+
+## Terminology / Concepts
+
+- **assertion** — a function that — given a value — if the value meets the assertion's criteria, quietly succeeds, but if the value fails the criteria, fails.
+- **CEL** — [Common Expression Language](https://github.com/google/cel-spec) — language supported (in alpha) by Kubernetes for CRD validation rules.
+- **module** — a file that can be loaded into a template in order to reuse its functions (or variables). The `ytt` "Standard Library" is comprised of modules that begin with the `@ytt:` prefix (e.g. @ytt:data, @ytt:assert, @ytt:overlay, ...) 
+- **node** — a single piece of YAML: a document set, document, map, map item, array, array item. One "annotates" a node by placing a `ytt` annotation-shaped comment just above the node (see also: [YAML Primer](https://carvel.dev/ytt/docs/v0.39.0/yaml-primer/)).
+- **required value** — a Data Value that the consumer must set. Mechanically, a Data Value that does not (yet) have a value that is "valid."
+- **rule** — the combination of: 
+    - a textual description of what constitutes a valid value;
+    - a function that asserts the rule against an actual value;
+    - a violation message template, used to report invalid values.
+    - **custom rule** — a rule based on a function (either one written by a user or provided in a `ytt` module).
+    - **named rule** — a rule provided with `ytt` expressed through a keyword argument to `@assert/validate`.
+- **validation** — the binding of one or more "rule"s to a "node".
+- **violation** — an instance of where a given value of a "node" fails a "validation".
+
+## Proposal
+
+Implement all plumbing logic required to provide high-quality data validation over Data Values and output documents.
+
+Mechanically, this means:
+
+Extend the `@ytt:assert` module, to include the ability to validate:
+- introduce an annotation (`@assert/validate`) which attaches one or more "rules" to a "node", defining what a valid value is for that node;
+- supply a set of common assertions (functions in the `@ytt:assert` module) that covers most validation use-cases;
+- provide guidance for how to express most commonly intended validations.
+
+Further, integrate this feature into `ytt`'s processing pipeline:
+- immediately after a final Data Values is calculated for a library, automatically validate that document (and halt processing if there are any violations);
+- if the user is inspecting schema, include validations in the output;
+- immediately after the Output Document Set (i.e. the final result of evaluating a library) is produced, automatically validate each document in the set (and halt processing if there are any violations).
+
+Finally, achieve all this in a way that complements existing validation tooling:
```

Comment:
Is this proposal a second layer of validation if a user doesn't opt into open api schema? I thought you could describe the validation rules through that mechanism?